[rllib] Correctly get bytes size of SampleBatch

RaphaelCS opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

To correctly get bytes size of `SampleBatch`.

## Related issue number

<!-- For example: "Closes #1234" -->

Closes #14800 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ericl commented:
Nice, seems some tests are failing (correctly) due to the check:
```
    def warn_replay_buffer_size(*, item: SampleBatchType, num_items: int) -> None:
        """Warn if the configured replay buffer size is too large."""
        if log_once("replay_buffer_size"):
            item_size = item.size_bytes()
            psutil_mem = psutil.virtual_memory()
            total_gb = psutil_mem.total / 1e9
            mem_size = num_items * item_size / 1e9
            msg = ("Estimated max memory usage for replay buffer is {} GB "
                   "({} batches of size {}, {} bytes each), "
                   "available system memory is {} GB".format(
                       mem_size, num_items, item.count, item_size, total_gb))
            if mem_size > total_gb:
>               raise ValueError(msg)
E               ValueError: Estimated max memory usage for replay buffer is 169.445 GB (1000000.0 batches of size 1, 169445 bytes each), available system memory is 8.34816 GB
```

Probably need to downsize the buffer for those.

ericl commented on rllib/policy/sample_batch.py:
```diff
@@ -426,7 +426,7 @@ def size_bytes(self) -> int:
             int: The overall size in bytes of the data buffer (all columns).
         """
-        return sum(sys.getsizeof(d) for d in self.values())
+        return sum(d.nbytes for d in self.data.values())
```
@sven1977 I think it's posible the values aren't numpy arrays right?

How about we use d.nbytes if isinstance(np.ndarray) otherwise, sys.getsizeof?

RaphaelCS replied:
If `d` isn't a numpy array, I think the best way is either 1) do nothing 2) throw a warning and do not estimate buffer size.

Actually `sys.getsizeof` is not guaranteed to get the real size of any non-primitive object.

In addition, as a user of rllib, I really suffered from the wrong estimation of memory usage. In my work, the estimation told me the buffer would take less than 1GB memory, but it actually would take dozens of GB. For this reason my program always went wrong and I saw some message like `the actor died unexpectedly`, which is difficult to locate the accurate bug.

sven1977 replied:
Correct, values dont have to be np arrays.
Can we add this check then? Also, please do not use `self.data` anymore. It's been deprecated as SampleBatches are now dicts.
So like this?
```
return sum(d.nbytes if isinstance(d, np.ndarray) else 0 for d in self.values())
```

ericl commented:
To avoid regressions could we keep the check anyways? I don't think there's
any harm and could catch non numpy cases.

On Fri, Mar 19, 2021, 7:34 PM Raphael CHEN ***@***.***> wrote:

> *@RaphaelCS* commented on this pull request.
> ------------------------------
>
> In rllib/policy/sample_batch.py
> <https://github.com/ray-project/ray/pull/14801#discussion_r598049678>:
>
> > @@ -426,7 +426,7 @@ def size_bytes(self) -> int:
>          Returns:
>              int: The overall size in bytes of the data buffer (all columns).
>          """
> -        return sum(sys.getsizeof(d) for d in self.values())
> +        return sum(d.nbytes for d in self.data.values())
>
> If d isn't a numpy array, I think the best way is either 1) do nothing 2)
> throw a warning and do not estimate buffer size.
>
> Actually sys.getsizeof is not guaranteed to get the real size of any
> non-primitive object.
>
> In addition, as a user of rllib, I really suffered from the wrong
> estimation of memory usage. In my work, the estimation told me the buffer
> would take less than 1GB memory, but it actually would take dozens of GB.
> For this reason my program always went wrong and I saw some message like the
> actor died unexpectedly, which is difficult to locate the accurate bug.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/14801#discussion_r598049678>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAADUSV7INJEDLKRBFE7AJ3TEQCUXANCNFSM4ZPBXKPA>
> .
>

RaphaelCS commented:
> To avoid regressions could we keep the check anyways? I don't think there's any harm and could catch non numpy cases.

Do you mean something like this?
(pseudo-code)
```
if d is np.adarray
    d.nbytes
else
    print "Non-numpy object may not result in a real bytes size!"
    sys.getsizeof(d)
```

sven1977 commented:
Thanks for the commit @RaphaelCS! Just one small change required to get this merged.

RaphaelCS commented:
@sven1977  I've changed it to deal with non-numpy object.

In addition, I've also added user warning when estimating maximum buffer size. I think it's really useful if we can't estimate real buffer size.
In fact, I think we should add user warn in method `size_bytes` since it is a `PublicAPI`. But I'm not sure whether it will show the warning too much time during training. (in `warn_replay_buffer_size` we are sure only `log_once`) Perhaps we can change `size_bytes` to `DeveloperAPI`?

BTW, I've wrongly used force-push so some conversation history is not valid. 😢

sven1977 reviewed (approved):
Looks good now. All RLlib test are passing. Thanks @RaphaelCS for the fixes!
